D2k - Implement original saboteur cloak logic

MustaphaTR opened the pull request:
Also removes ability to blow vehicles up from saboteur because that wasn't the case in original.

I did some tests in original game to find the values. On Ordos Test Mission you start with a preplaced Saboteur and a Palace. We know saboteur charges in 1 min 30 sec.
This image is when saboteur bar is full:
![resim](https://user-images.githubusercontent.com/7933210/46581787-e39c6880-ca46-11e8-9dd1-321a02f61620.png)
This is when bar is empty again, after i clicked deploy as soon as it charged: 
![resim](https://user-images.githubusercontent.com/7933210/46581793-fa42bf80-ca46-11e8-916e-46b5f8ebf27b.png)

So i concluded it charges in 50 sec and decharges in 15.

Closes #12542.
Closes #10083.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantTimedConditionOnDeploy.cs:
```diff
@@ -0,0 +1,262 @@
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class GrantTimedConditionOnDeployInfo : PausableConditionalTraitInfo
```
Can we please add time parameters to `GrantConditionOnDeploy` instead of adding whole new trait?

pchote commented on mods/d2k/rules/structures.yaml:
```diff
@@ -1155,7 +1155,7 @@ palace:
 	ProduceActorPower@saboteur:
 		Description: Recruit Saboteur
-		LongDesc: Sneaky infantry, armed with explosives\n  Strong vs Buildings\n  Weak vs Everything\n  Special Ability: destroy buildings
+		LongDesc: Sneaky infantry, armed with explosives.\nCan be deployed to become invisible for 15 seconds.\n  Strong vs Buildings\n  Weak vs Everything\n  Special Ability: destroy buildings
```
The specific time will depend on the game speed, so please don't mention it here.

MustaphaTR committed:
Add GrantConditionAfterDelay

MustaphaTR committed:
Add UndeployOnCondition to GrantConditionOnDeploy

MustaphaTR committed:
Add original Saboteur Cloak logic.

pchote commented on mods/d2k/rules/structures.yaml:
```diff
@@ -1173,7 +1173,7 @@ palace:
 	ProduceActorPower@saboteur:
 		Description: Recruit Saboteur
-		LongDesc: Sneaky infantry, armed with explosives\n  Strong vs Buildings\n  Weak vs Everything\n  Special Ability: destroy buildings
+		LongDesc: Sneaky infantry, armed with explosives.\nCan be deployed to become invisible for a while.\n  Strong vs Buildings\n  Weak vs Everything\n  Special Ability: destroy buildings
```
Please change this to something like "Can be deployed to become invisible for a limited time."

pchote commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -267,13 +267,26 @@ saboteur:
+		UncloakOn: Attack, Unload, Infiltrate, Demolish, Damage, Heal
 		IsPlayerPalette: true
 		PauseOnCondition: cloak-force-disabled
+		RequiresCondition: deployed
```
I'm not sure if this is still a controversial request, but IMO prefer `RequiresCondition` / `PauseOnCondition` at the top of trait definitions.

pchote commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -267,13 +267,26 @@ saboteur:
+		RequiresCondition: !deployed
+	GrantConditionAfterDelay@DECHARGE:
+		Delay: 375
+		Condition: decharged
```
English nit: use `discharged`.

pchote commented on mods/d2k/rules/infantry.yaml:
```diff
@@ -253,7 +253,7 @@ saboteur:
 		BuildPaletteOrder: 100
 		Prerequisites: ~disabled
-		Description: Sneaky infantry, armed with explosives\n  Strong vs Buildings\n  Weak vs Everything\n  Special Ability: destroy buildings
+		Description: Sneaky infantry, armed with explosives.\nCan be deployed to become invisible for a while.\n  Strong vs Buildings\n  Weak vs Everything\n  Special Ability: destroy buildings
```
Please change this to something like "Can be deployed to become invisible for a limited time."

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionAfterDelay.cs:
```diff
@@ -0,0 +1,103 @@
+
+		float ISelectionBar.GetValue()
+		{
+			if (IsTraitDisabled || !Info.ShowSelectionBar || ((1f - (float)Ticks / Info.Delay) > 1f && !info.ShowFullBarAfterGranted))
```
This line is a bit of a monster. Would you mind splitting this into one check for `IsTraitDisabled || !Info.ShowSelectionBar ` and then a second for the rest?

The `1f - (float)Ticks / Info.Delay) > 1f` looks dodgy, but my comment above applies here too.

MustaphaTR replied:
That ShowFullBarAfterGranted was a hack, timer shouldn't go down once the condition is granted as you said below.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionAfterDelay.cs:
```diff
@@ -0,0 +1,103 @@
+			if (IsTraitPaused || IsTraitDisabled)
+				return;
+
+			if (--Ticks < 0)
```
It looks like this will continue to count down without any lower bound.  This isn't good practice, and should be avoided.

I would need to invest some time into understanding exactly what this trait is trying to achieve before I could offer any specific advice for fixing this.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionAfterDelay.cs:
```diff
@@ -0,0 +1,103 @@
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Gives a condition to the actor after a delay.")]
+	public class GrantConditionAfterDelayInfo : PausableConditionalTraitInfo
```
I wonder whether it would make more sense to put an `InitialDelay` field on `GrantCondition` instead of having this as a separate trait.

pchote commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionAfterDelay.cs:
```diff
@@ -0,0 +1,103 @@
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Gives a condition to the actor after a delay.")]
+	public class GrantConditionAfterDelayInfo : PausableConditionalTraitInfo
```
I suspect that the best fix will be to invert the logic, and implement this as a `GrantTimedCondition` trait, or maybe via a `Duration` field on `GrantCondition` + the existing `TimedConditionBar` trait.

MustaphaTR replied:
I'm usnsure about this, i still think as we talked before getting a WithSelectionBarInfo with a bool to reverse it sounds like a good idea. Would be nice for unifying the traits around that. (We can do same with IPipInfo).

For complateness of this PR i can implement that only for this trait here, and we can adopt other traits later.

pchote commented:
Bumping to Next + 1.

MustaphaTR commented:
I don't think this should be a blocker for playtest either but, i mentioned this in the first issue of The Chrono so i would like to get this to the next release.

I'll try to fix the issues today.

MustaphaTR committed:
Add GrantTimedCondition

MustaphaTR commented:
Rebased and Updated. Current code crashed when the saboteur discharge is finished. I couldn't figure out how to solve it, so would be happy if someone can take a look.

MustaphaTR commented:
Rebased.

GraionDilach commented:
That sounds more in line with AS's GrantPeriodicConditionOnEvent with an additional Deploy event - https://github.com/AttacqueSuperior/Engine/blob/master/OpenRA.Mods.AS/Traits/Conditions/GrantPeriodicConditionOnEvent.cs.

pchote commented:
@MustaphaTR what do you want to do here?

reaperrr commented:
Since there hasn't been any progress here lately, I'll drop it off the milestone for now.

pchote commented:
Moving to the Future milestone to help clean up the PR queue.
